src/script/ismine.cpp

```diff
@@ -146,23 +144,27 @@ isminetype IsMineInner(const CKeyStore& keystore, const CScript& scriptPubKey, b
             }
         }
         if (HaveKeys(keys, keystore)) {
-            return ISMINE_SPENDABLE;
+            Update(ret, ISMINE_SPENDABLE);
         }
         break;
     }
     }
 
-    if (keystore.HaveWatchOnly(scriptPubKey)) {
+    if (ret == ISMINE_NO && keystore.HaveWatchOnly(scriptPubKey)) {
         return ISMINE_WATCH_ONLY;
     }
-    return ISMINE_NO;
+    return ret;
 }
 
 } // namespace
 
 isminetype IsMine(const CKeyStore& keystore, const CScript& scriptPubKey, bool& isInvalid)
 {
-    return IsMineInner(keystore, scriptPubKey, isInvalid, IsMineSigVersion::TOP);
+    isminetype ret = IsMineInner(keystore, scriptPubKey, isInvalid, IsMineSigVersion::TOP);
+    if (isInvalid) {
```

Comment:
commit: Simplify IsMine logic

It feels like `ISMINE_INVALID` should be an IsMine type. Would require some refactoring of the callers though...